X9.146/gen_dual_keysig_cert.c

```diff
@@ -0,0 +1,375 @@
+     * signature will be fatal. */
+    ret = wc_SignCert(newCert.bodySz, newCert.sigType, scratchBuf,
+                      scratchSz, &caKey, NULL, &rng);
+    if (ret < 0) goto exit;
+    printf("wc_SignCert for preTBS returned %d\n", ret);
+#endif
+    scratchSz = ret;
+
+    wc_InitDecodedCert(&preTBS, scratchBuf, scratchSz, 0);
+    initPreTBS = 1;
+    ret = wc_ParseCert(&preTBS, CERT_TYPE, NO_VERIFY, NULL);
+    if (ret < 0) goto exit;
+
+    /* Generate the DER for a pre-TBS. */
+    XMEMSET(preTbsBuf, 0, preTbsSz);
+    ret = GeneratePreTBS(&preTBS, preTbsBuf, preTbsSz);
+    if (ret < 0) goto exit;
+    printf("PreTBS is %d bytes.\n", ret);
+    preTbsSz = ret;
+
+    /* Generate the contents of the altSigVal extension and inject into cert. */
+    XMEMSET(altSigValBuf, 0, altSigValSz);
+    ret = wc_MakeSigWithBitStr(altSigValBuf, altSigValSz, CTC_SHA256wECDSA,
+                               preTbsBuf, preTbsSz, ECC_TYPE, &altCaKey, &rng);
+    if (ret < 0) goto exit;
+    altSigValSz = ret;
+    printf("altSigVal is %d bytes.\n", altSigValSz);
+
+    ret = wc_SetCustomExtension(&newCert, 0, "2.5.29.74",
+                                altSigValBuf, altSigValSz);
+    if (ret < 0) goto exit;
+
+    /* Finally, generate the new certificate. */
+    XMEMSET(outBuf, 0, outSz);
+#ifdef GEN_ROOT_CERT
+    ret = wc_MakeSelfCert(&newCert, outBuf, outSz, &caKey, &rng);
+    if (ret <= 0) goto exit;
+    printf("wc_MakeSelfCert for preTBS returned %d\n", ret);
+#else
+    ret = wc_MakeCert(&newCert, outBuf, outSz, &serverKey, NULL, &rng);
+    if (ret < 0) goto exit;
+    printf("Make Cert returned %d\n", ret);
+
+    ret = wc_SignCert(newCert.bodySz, newCert.sigType, outBuf, outSz, &caKey,
+                      NULL, &rng);
+    if (ret < 0) goto exit;
+    printf("Sign Cert returned %d\n", ret);
+#endif
+    outSz = ret;
+
+    printf("Successfully created new certificate\n\n");
+
+    /* Write the new cert to file in der format. */
+    printf("Writing newly generated DER certificate to file \"%s\"\n",
+           newCertOutput);
+    file = fopen(newCertOutput, "wb");
+    if (!file) {
+        printf("failed to open file: %s\n", newCertOutput);
+        goto exit;
+    }
+    ret = fwrite(outBuf, 1, outSz, file);
+    fclose(file);
+    if (ret <= 0) goto exit;
+    printf("Successfully output %d bytes\n", ret);
+
+    ret = 0;
+    printf("SUCCESS!\n");
+exit:
+
+    if (initCaKey)
+        wc_FreeRsaKey(&caKey);
+#ifndef GEN_ROOT_CERT
+    if (initServerKey)
+        wc_FreeRsaKey(&serverKey);
+#endif
+    if (initPreTBS)
+        wc_FreeDecodedCert(&preTBS);
+    if (initRng)
+        wc_FreeRng(&rng);
+
+    if (ret != 0)
+        printf("Failure code was %d\n", ret);
+    return ret;
+}
+
+int main(int argc, char** argv)
+{
+    return do_certgen(argc, argv);
+}
+
+#else
+
+int main(int argc, char** argv)
+{
+    printf("Please compile wolfSSL with --enable-x9-146 or "

```

Comment:
Are these enable options in the examples still accurate? The PR on wolfssl/master referenced adds --enable-dual-alg-certs.

Reply:
Fixed.  Good eye @JacobBarthelmeh !